python/dgl/node_flow.py

```diff
@@ -0,0 +1,845 @@
+        block_id : int or ALL
+            the block to register the message function.
+        """
+        if is_all(block_id):
+            self._message_funcs = [func] * self.num_blocks
+        else:
+            self._message_funcs[block_id] = func
+
+    def register_reduce_func(self, func, block_id=ALL):
+        """Register global message reduce function for a block.
+
+        Once registered, ``func`` will be used as the default
+        message reduce function in message passing operations, including
+        :func:`block_compute`, :func:`prop_flow`.
+
+        Parameters
+        ----------
+        func : callable
+            Reduce function on the node. The function should be
+            a :mod:`Node UDF <dgl.udf>`.
+        block_id : int or ALL
+            the block to register the reduce function.
+        """
+        if is_all(block_id):
+            self._reduce_funcs = [func] * self.num_blocks
+        else:
+            self._reduce_funcs[block_id] = func
+
+    def register_apply_node_func(self, func, block_id=ALL):
+        """Register global node apply function for a block.
+
+        Once registered, ``func`` will be used as the default apply
+        node function. Related operations include :func:`apply_layer`,
+        :func:`block_compute`, :func:`prop_flow`.
+
+        Parameters
+        ----------
+        func : callable
+            Apply function on the nodes. The function should be
+            a :mod:`Node UDF <dgl.udf>`.
+        block_id : int or ALL
+            the block to register the apply node function.
+        """
+        if is_all(block_id):
+            self._apply_node_funcs = [func] * self.num_blocks
+        else:
+            self._apply_node_funcs[block_id] = func
+
+    def register_apply_edge_func(self, func, block_id=ALL):
+        """Register global edge apply function for a block.
+
+        Once registered, ``func`` will be used as the default apply
+        edge function in :func:`apply_block`.
+
+        Parameters
+        ----------
+        func : callable
+            Apply function on the edge. The function should be
+            an :mod:`Edge UDF <dgl.udf>`.
+        block_id : int or ALL
+            the block to register the apply edge function.
+        """
+        if is_all(block_id):
+            self._apply_edge_funcs = [func] * self.num_blocks
+        else:
+            self._apply_edge_funcs[block_id] = func
+
+    def apply_layer(self, layer_id, func="default", v=ALL, inplace=False):
+        """Apply node update function on the node embeddings in the specified layer.
+
+        Parameters
+        ----------
+        layer_id : int
+            The specified layer to update node embeddings.
+        func : callable or None, optional
+            Apply function on the nodes. The function should be
+            a :mod:`Node UDF <dgl.udf>`.
+        v : a list of vertex Ids or ALL.
+            The vertices to run the node update function.
```

Comment:
Is `v` local or global?

Reply:
what is the definition of local and global here?

Reply:
local is the node id of each layer; global is the id of the whole graph

Reply:
i didn't allow node id of each layer. it's kind of confusing. there would be parent vertex id space, nodeflow vertex id space and layer id space.